pkg/kubelet/oom/oom_watcher_linux_test.go

```diff
@@ -26,19 +28,152 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-// TestBasic verifies that the OOMWatch works without error.
-func TestBasic(t *testing.T) {
+type fakeStreamer struct {
+	oomInstancesToStream []*oomparser.OomInstance
+}
+
+func (fs *fakeStreamer) StreamOoms(outStream chan<- *oomparser.OomInstance) {
+	for _, oomInstance := range fs.oomInstancesToStream {
+		outStream <- oomInstance
+	}
+}
+
+// TestStartingWatcher tests that the watcher, using the actual streamer
+// and not the fake, starts successfully.
+func TestStartingWatcher(t *testing.T) {
 	fakeRecorder := &record.FakeRecorder{}
 	node := &v1.ObjectReference{}
 
-	// TODO: Substitute this `oomStreamer` out for a fake, and then write
-	// more comprehensive unit tests of the actual behavior.
-	oomStreamer, err := oomparser.New()
+	oomWatcher, err := NewWatcher(fakeRecorder)
 	assert.NoError(t, err)
+	assert.NoError(t, oomWatcher.Start(node))
+}
+
+// TestWatcherRecordsEventsForOomEvents ensures that our OomInstances coming
+// from `StreamOoms` are translated into events in our recorder.
+func TestWatcherRecordsEventsForOomEvents(t *testing.T) {
+	oomInstancesToStream := []*oomparser.OomInstance{
+		{
+			Pid:                 1000,
+			ProcessName:         "fakeProcess",
+			TimeOfDeath:         time.Now(),
+			ContainerName:       recordEventContainerName,
+			VictimContainerName: "some-container",
+		},
+	}
+	numExpectedOomEvents := len(oomInstancesToStream)
+
+	fakeStreamer := &fakeStreamer{
+		oomInstancesToStream: oomInstancesToStream,
+	}
+
+	fakeRecorder := record.NewFakeRecorder(numExpectedOomEvents)
+	node := &v1.ObjectReference{}
+
+	oomWatcher := &realWatcher{
+		recorder:    fakeRecorder,
+		oomStreamer: fakeStreamer,
+	}
+	assert.NoError(t, oomWatcher.Start(node))
+
+	eventsRecorded := getRecordedEvents(fakeRecorder, numExpectedOomEvents)
+	assert.Equal(t, numExpectedOomEvents, len(eventsRecorded))
+}
+
+func getRecordedEvents(fakeRecorder *record.FakeRecorder, numExpectedOomEvents int) []string {
+	eventsRecorded := []string{}
+
+	select {
+	case event := <-fakeRecorder.Events:
+		eventsRecorded = append(eventsRecorded, event)
+
+		if len(eventsRecorded) == numExpectedOomEvents {
+			break
+		}
+	case <-time.After(10 * time.Second):
```

Comment:
Can you run `go test` with `--race` on this file?  I have a vague recollection that anything with a timeout doesn't pass the race detector.

If it fails, the alternative is to have this function block until it receives numExpectedOomEvents.  While it is annoying to wait 45s for the test to timeout on failure, it is good to write tests that pass the race checker.